### DISCUSSION TITLE
An API skeleton

### DISCUSSION THREAD

**Discussion by hose1021** [laravel/framework#39514] [Ideas] [answered]
An API skeleton, generated via something like:

laravel new project-orange —api

which will remove all files related to views, assets, laravel mix, package.json and other similar files that are not used in an API only project.

Link to [original](https://twitter.com/davorminchorov/status/1456658503146299395) (c) @davorminchorov

**Comment by davorminchorov** [comment]
Here's the list of changes I am thinking of:

```
	deleted:    resources/css/app.css
	deleted:    resources/js/app.js
	deleted:    resources/js/bootstrap.js
	deleted:    resources/views/welcome.blade.php
	modified:   routes/api.php
	modified:   routes/web.php
	deleted:    webpack.mix.js
```
In the `routes/web.php`, I replaced the default route to a redirect to `/api`:
```
Route::redirect('/', '/api');
```
and I updated the `routes/api.php` file with the following new route:
```
Route::get('/', function() {
    return response()->json([
        'name' => config('app.name'),
        'version' => [
            'laravel' => Illuminate\Foundation\Application::VERSION,
            'php' => PHP_VERSION,
        ],
        'features' => [
            [
                'name' => 'Documentation',
                'description' => 'Laravel has wonderful, thorough documentation covering every aspect of the framework. Whether you are new to the framework or have previous experience with Laravel, we recommend reading all of the documentation from beginning to end.',
                'url' => 'https://laravel.com/docs',
            ],
        ],
        'shop_url' => 'https://laravel.bigcartel.com',
        'sponsor_url' => 'https://github.com/sponsors/taylorotwell',
    ]);
});
```
It's just an example but It could be updated with all of the values that the default `welcome.blade.php` file has for the first time API skeleton setup, so that the developers can know that everything works fine when they run the script and open the app in the browser.

I also created a `resources/views/emails` folder just to demonstrate that the developer will need the views folder in case his/her API sends emails.

If anyone has any other ideas about changes, feel free to share them.

**Comment by dennisprudlo** [comment]
I wouldn't redirect the '/' web route to the api. Say you use `example.com/api` as your api and some React app displays a landing page on `example.com`. Then your api-app and react-app will fight over what happens when `example.com` ist requested.

If your api is at `example.com/api` then people should call it at that endpoint. I think getting completely rid of the `web.php` and updating the routing in the `RouteServiceProvider` would remove more "unused code" for an api-only application.

```php
$this->routes(function () {
    Route::prefix('api')->middleware('api')
        ->namespace($this->namespace)
        ->group(base_path('routes/api.php'));

    Route::middleware('web')
        ->namespace($this->namespace)
        ->group(base_path('routes/web.php'));
});
```
This can be reduced to:
```php
$this->routes(function () {
    Route::middleware('api')
        ->namespace($this->namespace)
        ->group(base_path('routes/api.php'));
});
```
Also getting rid of the `/api`-prefix so `/` is the new api root.

**Reply by davorminchorov** [reply]
The reason I left the web route is that it's a starting point for the developer to see if everything works fine after the initial install. 

There are different ways to architect an API based on different developer choices and project requirements.

The default skeleton cannot predict the choices of the developers how they will develop the application further after the initial setup.

**Reply by dennisprudlo** [reply]
Do you mean that in the future the developer may decide to add web views to the application? Wouldn't that then also apply to removing the css and js files in the resources?

**Reply by davorminchorov** [reply]
My idea was that if you are going to build an API, you most likely wouldn't need any of these things and most likely develop the frontend app (react / vue / swift / kotlin) in a separate repository.

If you were to need the removed files, you wouldn't use the --api flag while installing the project.

**Reply by dennisprudlo** [reply]
Yes that what I meant. If someone used the --api flag I would assume they don't need the web routes preconfigured. That's why I proposed to remove them (just like the resources).

**Reply by davorminchorov** [reply]
I get the idea and I agree, but it might be confusing for beginners and/or inexperienced people.

I am curious to see what other people think and we might keep it or remove it based on a final decision.

**Reply by hose1021** [reply]
I think web.php must be. For example, api.php after install have route with sanctum. Also need to have web.php, but just need to change comment for `'/'`.

P.S. sorry about my bad english.

**Comment by davorminchorov** [comment, marked as answer]
[Taylor is already working on the solution](https://twitter.com/taylorotwell/status/1457829364163387392?s=21).

**Comment by scottgc** [comment]
Is there a branch or issue that work has begun on this idea that we can begin to contribute to.
